File: source/cpp/modulators.cpp
#include "modulators.h"

#include <stdlib.h>
#include <iostream>                                                             
#include <ctime>                                                                
#include <string>                                                               
#include <random>                                                               
#include <cmath>                                                                
#include <complex>                                                              


#define PI 3.14159265

using namespace std;
using std::string;
//string::size_type sz;

vector<string> chunkBits(int bitspersymbol, string &input, char pad)
{
    int n_chunks = input.length()/bitspersymbol;
    vector<string> output;
    string res = "";

    for (int i=0; i<n_chunks; i++)
    {
        for (int j=0; j<bitspersymbol; j++)
        {
            res += input[i*bitspersymbol+j];
            if (res.length() == bitspersymbol)
            {
                output.push_back(res);
                res = "";
            }
        }
    }

    if (res != "")
    {
        while (res.length() < bitspersymbol)
        {
            res += pad;
        }
        output.push_back(res);
    }
    
    return output;
}

Modulator::Modulator(){};

void Modulator::setOrder(int O)
{
    mod_order = O;
    bitspersymbol = sqrt(mod_order);

}

ModulatorBPSK::ModulatorBPSK(){
    mod_order = 2;
    bitspersymbol = 1;
};

//class ModulatorBPSK : public Modulator
complex<double>* ModulatorBPSK::modulate(string &input)
{
    const char* c_bit = input.c_str();

    int data_len = input.length();
    complex<double>* output = (complex<double>*) malloc(data_len*sizeof(complex<double>));

    for (int i=0; i<data_len; i++)
    {
        int i_bit = c_bit[i] - '0'; 
        output[i] = cos(PI*(1 - i_bit));
    }

    return output;
};

//class ModulatorQPSK : public Modulator
ModulatorQPSK::ModulatorQPSK()
{
    mod_order = 4;
    bitspersymbol = 2;
}

complex<double>* ModulatorQPSK::modulate(string &input)
{
    vector<string> saneInput = chunkBits(2, input, '0');

    int output_len = saneInput.size();

    complex<double>* output = (complex<double>*) malloc(output_len*sizeof(complex<double>));

    for (int i=0; i<output_len; i++)
    {
        int i_bit = stoi(saneInput[i], nullptr, 2); 
        double I_sample = cos((PI/4)*(2*i_bit + 1));
        double Q_sample = sin((PI/4)*(2*i_bit + 1));
        complex<double> sample(I_sample, Q_sample);
        output[i] = sample;
    }

    return output;
}



//class ModulatorMPSK : public Modulator
ModulatorMPSK::ModulatorMPSK(){};

complex<double>* ModulatorMPSK::modulate(string &input)
{
    vector<string> saneInput = chunkBits(bitspersymbol, input, '0');

    int output_len = saneInput.size();

    complex<double>* output = (complex<double>*) malloc(output_len*sizeof(complex<double>));

    for (int i=0; i<output_len; i++)
    {
        int i_bit = stoi(saneInput[i], nullptr, 2); 
        double I_sample = cos((2*PI/mod_order)*(i_bit));
        double Q_sample = sin((2*PI/mod_order)*(i_bit));
        complex<double> sample(I_sample, Q_sample);
        output[i] = sample;
    }

    return output;
}


/*
int main()
{
    string data = "0001101100011011";
    int mod_order = 4;

    ModulatorQPSK modulator;
    modulator.setOrder(mod_order);

    complex<double> * output = modulator.modulate(data);

    int tx_symbols = data.length()/sqrt(mod_order);
    
    for (int i=0; i < tx_symbols; i++)
    {
        cout<<output[i]<<endl;
    }

    return 0;
}*/

File: source/python/modulators.py
#! /usr/bin/env python3
# -*- docing : utf8 -*-

import numpy as np
from abc import ABC, abstractmethod

from utils import base_2_to_10_array, binaryToGray
from channel import AwgnRayleighChannel
from demodulators import DemodulatorBPSK
from detectors import ZeroForcing

__all__ = ["Modulator","ModulatorBPSK", "ModulatorQPSK", "ModulatorMPSK"]

class Modulator:
    def __init__(self, order):
        self.mod_order = order
        self.bits_per_symbol = np.log2(order);

    def modulate(self, data_input):
        pass


class ModulatorBPSK(Modulator):
    def __init__(self):
        order = 2
        super().__init__(order)
        #self.mod_order = 2
        #self.bits_per_symbol = 1

    def modulate(self, data_input):
        #Input must always be a tuple of 0 or 1, no matter the type
        modulated = []
        for bit in data_input:
            modulated.append((2*int(bit) - 1) + 0j)

        return modulated

class ModulatorQPSK(Modulator):
    def __init__(self):
        order = 4
        super().__init__(order)
        #self.mod_order = 4
        #self.bits_per_symbol = 2

    def modulate(self, data_input):
        modulated = []
        chunk = []
        for bit in data_input:
            chunk.append(bit)
            if len(chunk) == 2:
                symb = base_2_to_10_array(chunk)
                i_sample = np.cos((np.pi/4)*(2*symb + 1))
                q_sample = np.sin((np.pi/4)*(2*symb + 1))

                modulated.append(i_sample + q_sample*1j)
                chunk = []
        
        return modulated


class ModulatorMPSK(Modulator):
    def __init__(self, order):
        super().__init__(order)
        #self.mod_order = order
        #self.bits_per_symbol = np.log2(order)

    def modulate(self, data_input):
        if self.mod_order == 0:
            raise AttributeError('Modulation order not especified!')

        self.bits_per_symbol = np.log2(self.mod_order)

        modulated = []
        chunk = []
        for bit in data_input:
            chunk += bit
            if len(chunk) == self.bits_per_symbol:
                symb = base_2_to_10_array(chunk)
                i_sample = np.cos((np.pi/self.mod_order)*(symb))
                q_sample = np.sin((np.pi/self.mod_order)*(symb))
                modulated.append(i_sample + q_sample*1j)

                chunk = []
        
        return modulated

class ModulatorMQAM(Modulator):
    '''Based on the python code for QAM modulator found at:
    https://www.gaussianwaves.com/2012/10/qam-modulation-simulation-matlab-python/
    '''
    def __init__(self, order):
        super().__init__(order)
        self.graycode = True

    def modulate(self, data_input):
        if self.mod_order == 0:
            raise AttributeError('Modulation order not especified!')

        self.bits_per_symbol = np.log2(self.mod_order)
        grid_dimension = np.sqrt(self.mod_order).astype(int)

        modulated = []
        chunk = '' 
        for bit in data_input:
            chunk += bit
            if len(chunk) == self.bits_per_symbol:
                if self.graycode:
                    chunk = binaryToGray(chunk)

                symb = base_2_to_10_array(chunk)

                (x, y) = np.divmod(symb, grid_dimension)
                i_sample = 2*x+1-grid_dimension
                q_sample = 2*y+1-grid_dimension

                modulated.append(i_sample + q_sample*1j)

                chunk = ''

        return modulated

    def constellation(self):
        stream = ''

        for num in range(self.mod_order):
            stream += np.binary_repr(num,int(self.bits_per_symbol))

        return self.modulate(stream)

